### DISCUSSION TITLE
what's the function of INITIALIZER in proxy

### DISCUSSION THREAD

**Discussion by khushbusandhu** [Cyfrin/foundry-full-course-cu#709] [General]
as Patrick said proxies don't use a constructor but implementation contracts do and we have first to call the proxy that directs us to the implementation function. so in this process why do we need to have an initializer since the constructor will be called in the implementation contract
And what does it mean that storage is stored in proxy, not in implementation, how is it linked with Initializer

**Comment by theirrationalone** [comment]
Sorry to say, But still learning about it, Can't say anything yet!

**Reply by khushbusandhu** [reply]
okay no problem 
I am also searching for it

**Reply by khushbusandhu** [reply]
hey @theirrationalone can you tell me any resource from where I can learn about this proxy stuff?

**Reply by theirrationalone** [reply]
Johnny Times, Youtuber, 

https://youtu.be/JgSj7IiE4jA?si=_SUzZQHPmTzcwmpo

Openzeppelin.... , 


https://dev.to/yakult/tutorial-write-upgradeable-smart-contract-proxy-contract-with-openzeppelin-1916

**Reply by khushbusandhu** [reply]
okay thanks @theirrationalone

**Comment by AnouarBF** [comment]
Well What I understand is that the purpose of the proxy is to help upgrade our smart contracts. When storage is stored at proxy because the proxy stays there and from its address where users interact with our protocol. When our users wants to store a number for example this proxy calls our smart contract which has the setNumber function logic and use it to store a number in its storage. Therefore we don't need to change the whole system in case there is bug or a new feature released, we only need to change the contract that has the logic or features (That contract is called implementation). Now Openzeppelin helps us by giving a prepared proxy, since the only thing they do is performing calls to the correct implementation and storing results and data init. 
Now Think about it, If we already have a proxy that stores all of our data and users interact with it only. Why do we need a constructor in our implementation? it only holds functions, error handlers, modifiers to perform certain logic. we don't want to store anything in any implementation, since they could be changed (upgraded) later, and that could lead to data loss.
Hope that was helpful, I might have some confusions or mistakes so keep learning and asking.

**Reply by khushbusandhu** [reply]
Thanks a lot, @AnouarBF It helped me clear some of my doubts 
but I have to work more on the basics and I am working on it , thanks for sharing your knowledge
